src/librbd/crypto/LoadRequest.cc

```diff
@@ -46,20 +57,78 @@ void LoadRequest<I>::send() {
     ictx = ictx->parent;
   }
 
+  m_current_image_ctx = m_image_ctx;
+  load();
+}
+
+template <typename I>
+void LoadRequest<I>::load() {
+  m_detected_format_name = "";
+  ldout(m_image_ctx->cct, 20) << "format_idx=" << m_format_idx << dendl;
   auto ctx = create_context_callback<
-          LoadRequest<I>, &LoadRequest<I>::finish>(this);
-  m_format->load(m_image_ctx, ctx);
+          LoadRequest<I>, &LoadRequest<I>::handle_load>(this);
+  m_formats[m_format_idx]->load(m_current_image_ctx, &m_detected_format_name,
+                                ctx);
+}
+
+template <typename I>
+void LoadRequest<I>::handle_load(int r) {
+  ldout(m_image_ctx->cct, 20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    if (m_is_current_format_cloned && m_detected_format_name == "unknown") {
+      // encryption format was not detected, assume plaintext
+      ldout(m_image_ctx->cct, 5) << "assuming plaintext for image "
+                                 << m_current_image_ctx->name << dendl;
+      m_formats.pop_back();
+      finish(0);
+      return;
+    }
+
+    lderr(m_image_ctx->cct) << "failed to load encryption. image name: "
+                            << m_current_image_ctx->name << dendl;
+    finish(r);
+    return;
+  }
+
+  ldout(m_image_ctx->cct, 5) << "loaded format " << m_detected_format_name
+                             << (m_is_current_format_cloned ? " (cloned)" : "")
+                             << " for image " << m_current_image_ctx->name
+                             << dendl;
+
+  m_format_idx++;
+  m_current_image_ctx = m_current_image_ctx->parent;
+  if (m_current_image_ctx != nullptr) {
+    // move on to loading parent
+    if (m_format_idx >= m_formats.size()) {
+      // try to load next ancestor using the same format
+      ldout(m_image_ctx->cct, 20) << "cloning format" << dendl;
+      m_formats.push_back(m_formats[m_formats.size() - 1]->clone());
+      m_is_current_format_cloned = true;
+    }
+
+    load();
+  } else {
+    if (m_formats.size() > m_format_idx) {
+      lderr(m_image_ctx->cct) << "got " << m_formats.size()
+                              << " encryption specs to load, "
+                              << "but image has " << m_format_idx - 1
+                              << " ancestors" << dendl;
+      r = -EINVAL;
+    }
+
+    finish(r);
+  }
 }
 
 template <typename I>
 void LoadRequest<I>::finish(int r) {
+  ldout(m_image_ctx->cct, 20) << "r=" << r << dendl;
 
   if (r == 0) {
-    // load crypto layers to image and its ancestors
-    auto crypto = m_format->get_crypto();
     auto ictx = m_image_ctx;
-    while (ictx != nullptr) {
-      util::set_crypto(ictx, crypto);
+    for (auto& format : m_formats) {
+      util::set_crypto(ictx, std::move(format));
```

Comment:
`TestCloneEncryption` fails if persistent write log cache is enabled:
```
[ RUN      ] TestLibRBD.TestCloneEncryption
using new format!
wrote: 4
wrote: 4
wrote: 4
read: 4
read: RBDL
expected: aaaa
/build/ceph-17.0.0-13779-gbc834a42/src/test/librbd/test_librbd.cc:365: Failure
Expected equality of these values:
  0
  memcmp(result, expected, len)
    Which is: -1
read: 4
read: 4
read: 4
read: 4
read: 4
[  FAILED  ] TestLibRBD.TestCloneEncryption (49933 ms)
```
This is because "raw" reads performed by `rbd_encryption_load2` are cached.  The same can probably occur with the regular in-memory cache which is on by default.  I think the load operation should flush and invalidate the cache before returning, similar to how it's done on resize.

Reply:
Where do you think it's best to do this flush and invalidate? in `crypto/LoadRequest.cc` or `crypto/luks/LoadRequest.cc`?

Reply:
Probably in `crypto/LoadRequest.cc` since flush and invalidate would be needed for a hypothetical non-LUKS format as well.